File: backend/src/main/java/com/apartmentmanagement/entity/Notification.java
package com.apartmentmanagement.entity;

import java.time.LocalDateTime;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import com.apartmentmanagement.enums.NotificationType;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "notifications")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Notification {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String content;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private NotificationType type;

    /**
     * Tên bảng liên quan để deep link: "bills", "service_requests", "vehicles",...
     * NULL = thông báo không liên quan tới record cụ thể
     */
    @Column(name = "reference_type", length = 30)
    private String referenceType;

    @Column(name = "reference_id")
    private Long referenceId;

    @Column(name = "is_read", nullable = false)
    @Builder.Default
    private Boolean isRead = false;

    /** NULL cho tới khi người dùng đọc */
    @Column(name = "read_at")
    private LocalDateTime readAt;

    // --Relationships--

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // --Audit--

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    // --Helper--

    public void markAsRead() {
        this.isRead = true;
        this.readAt = LocalDateTime.now();
    }
}

File: frontend/src/types/notification.ts
import type { AnnouncementPriority, NotificationType } from "./common";

export interface Notification {
  id: number;
  title: string;
  content: string | null;
  type: NotificationType;
  referenceType: string | null;
  referenceId: number | null;
  isRead: boolean;
  readAt: string | null;
  createdAt: string;
}

export interface Announcement {
  id: number;
  title: string;
  content: string;
  priority: AnnouncementPriority | null;
  senderName: string;
  buildingId?: number;
  attachmentUrls: string[];
  isPublished: boolean;
  publishedAt: string;
  expiresAt: string | null;
  viewCount?: number;
  createdAt: string;
}

export interface AnnouncementCreateRequest {
  title: string;
  content: string;
  priority: AnnouncementPriority;
  attachmentUrls?: string[];
  expiresAt?: string;
}
